examples/sha256_nivc_lem.rs

```diff
@@ -0,0 +1,133 @@
+                    (if (eq 'lurk type)
+                        (commit value)
+                        (if (eq 'id type)
+                            value))))))
+      (encode (lambda (input)
+                (if input
+                    (cons 
+                        (encode-1 (car input))
+                        (encode (cdr input)))))))
+  (encode '((sha256 . {input}))))
+"#
+    );
+
+    store.read_with_default_state(&program).unwrap()
+}
+
+/// Run the example in this file with
+/// `cargo run --release --example sha256_nivc_lem <n>`
+/// where `n` is the needed arity (default is 1)
+fn main() {
+    let subscriber = Registry::default()
+        .with(fmt::layer().pretty())
+        .with(EnvFilter::from_default_env())
+        .with(TeXRayLayer::new());
+    tracing::subscriber::set_global_default(subscriber).unwrap();
+
+    let args = std::env::args().collect::<Vec<_>>();
+    let n = args.get(1).unwrap_or(&"1".into()).parse().unwrap();
+
+    let store = &Store::<Fr>::default();
+    let cproc_sym = user_sym(&format!("sha256_nivc_{n}"));
+
+    let call = sha256_nivc(store, n, (0..n).collect());
+
+    let mut lang = Lang::<Fr, Sha256Coprocessor<Fr>>::new();
+    lang.add_coprocessor_lem(cproc_sym, Sha256Coprocessor::new(n), store);
+    let lang_rc = Arc::new(lang.clone());
+
+    let lurk_step = make_eval_step_from_lang(&lang, false);
+    let (frames, _) = evaluate(Some((&lurk_step, &lang)), call, store, 1000).unwrap();
+
+    let supernova_prover = SuperNovaProver::<Fr, Sha256Coprocessor<Fr>, MultiFrame<'_, _, _>>::new(
+        REDUCTION_COUNT,
+        lang,
+    );
+
+    println!("Setting up running claim parameters (rc = {REDUCTION_COUNT})...");
+    let pp_start = Instant::now();
+
+    let instance_primary = Instance::new(
+        REDUCTION_COUNT,
+        lang_rc.clone(),
+        true,
+        Kind::SuperNovaAuxParams,
+    );
+    let pp = supernova_public_params::<_, _, MultiFrame<'_, _, _>>(
+        &instance_primary,
+        &public_params_default_dir(),
+    )
+    .unwrap();
+
+    let pp_end = pp_start.elapsed();
+    println!("Running claim parameters took {:?}", pp_end);
+
+    println!("Beginning proof step...");
+    let proof_start = Instant::now();
+    let (proof, z0, zi, num_steps, last_circuit_index) =
+        tracing_texray::examine(tracing::info_span!("bang!")).in_scope(|| {
+            supernova_prover
+                .prove(&pp, &frames, store, lang_rc)
+                .unwrap()
+        });
+    let proof_end = proof_start.elapsed();
+
+    println!("Proofs took {:?}", proof_end);
+
+    println!("Verifying proof...");
+
+    let verify_start = Instant::now();
+    let res = proof
+        .verify(&pp, last_circuit_index, num_steps, &z0, &zi)
+        .unwrap();
+    let verify_end = verify_start.elapsed();
+
+    println!("Verify took {:?}", verify_end);
+
+    if res {
+        println!(
+            "Congratulations! You proved and verified a SHA256 hash calculation in {:?} time!",
```

Comment:
Nit: it would be nice if I could know if this is a LEM calculation just by looking at the printout. IOW I would qualify "verified a LEM NIVC SHA256..." here and elsewhere.